Gutenboarding: design picker animations with react-spring

razvanpapadopol opened the pull request:
#### Changes proposed in this Pull Request

* Replace `react-transition-group` and CSS transitions with `react-spring`

#### Testing instructions
* Go to /gutenboarding/design
* When selecting a design the animations should follow the existing pattern but feel more natural

Fixes https://github.com/Automattic/wp-calypso/issues/38612

Commit:
Cleanup design picker animations and introduce react-spring.
- remove isSelected prop from DesignCard since it's not used
- remove cp from state and set position within grid for selected card
- use react-spring to handle design-selector animation

Commit:
Use react-spring for design-selector__description-container

Commit:
Use react-spring for page-layout-container
* clean up CSS variables related to animation
* remove isDialogVisible state
* remove transitionTiming const

simison commented:
FYI asked Ganon team for review just in case anyone from there would like to learn about react-spring while reviewing. :-)

Commit:
Fix page-selector fade-out animation.
- manage visibility of the Dialog with  useDialogState
- use onStart and onRest callbacks of react-spring to show and hide Dialog before and after animation

p-jackson commented on client/landing/gutenboarding/onboarding-block/design-selector/index.tsx:
```diff
@@ -67,20 +64,35 @@ const DesignSelector: FunctionComponent = () => {
+	const designSelectorSpring = useSpring( {
+		transform: `translate3d( 0, ${
+			hasSelectedDesign ? -selectionTransitionShift.current : 0
+		}px, 0)`,
```
nitpick

```suggestion
		}px, 0 )`,
```

p-jackson commented on client/landing/gutenboarding/onboarding-block/design-selector/index.tsx:
```diff
@@ -67,20 +64,35 @@ const DesignSelector: FunctionComponent = () => {
+	} );
+
+	const descriptionContainerSpring = useSpring( {
+		transform: `translate3d( 0, ${ hasSelectedDesign ? '0' : '100vh' }, 0)`,
```
nitpick

```suggestion
		transform: `translate3d( 0, ${ hasSelectedDesign ? '0' : '100vh' }, 0 )`,
```

p-jackson commented on client/landing/gutenboarding/onboarding-block/design-selector/index.tsx:
```diff
@@ -67,20 +64,35 @@ const DesignSelector: FunctionComponent = () => {
+	} );
+
+	const pageSelectorSpring = useSpring( {
+		transform: `translate3d( 0, ${ hasSelectedDesign ? '0' : '100vh' }, 0)`,
```
nitpick

```suggestion
		transform: `translate3d( 0, ${ hasSelectedDesign ? '0' : '100vh' }, 0 )`,
```

p-jackson reviewed (approved):
Looks great, much better animations on FF!

The docs said that using the "View interpolation" method can be a little faster sometimes; but I gave it a go and didn't see a noticable difference in the profiler. I think doing interoplation in the spring looks nicer so I'd stick with what you've got :)

Transition is broken in Safari, but that's an existing issue.

:shipit:

p-jackson commented:
Actually another point @razvanpapadopol, I would have expected to see a `package-lock.json` change in this PR with the addition of `react-spring`?

Commit:
Update client/landing/gutenboarding/onboarding-block/design-selector/index.tsx

Co-Authored-By: Philip Jackson <p-jackson@live.com>

Commit:
Update client/landing/gutenboarding/onboarding-block/design-selector/index.tsx

Co-Authored-By: Philip Jackson <p-jackson@live.com>

Commit:
Update client/landing/gutenboarding/onboarding-block/design-selector/index.tsx

Co-Authored-By: Philip Jackson <p-jackson@live.com>

razvanpapadopol commented:
> I would have expected to see a `package-lock.json` change in this PR with the addition of `react-spring`?

Yeah, @sirreal mentioned this too in the [other review](https://github.com/Automattic/wp-calypso/pull/38721/files#r364742817) but it seems it's already there (added 5 months ago) with the exact same version. Also I can see it's a dependency for `@wordpress/block-editor` and `@wordpress/components`.

sirreal commented:
> > I would have expected to see a `package-lock.json` change in this PR with the addition of `react-spring`?
> 
> Yeah, @sirreal mentioned this too in the [other review](https://github.com/Automattic/wp-calypso/pull/38721/files#r364742817) but it seems it's already there (added 5 months ago) with the exact same version. Also I can see it's a dependency for `@wordpress/block-editor` and `@wordpress/components`.

It does seem to be in there, it's likely been deduped/hoisted as a shared dep.

This will need a rebase after the `client` reorganization.
